File: Server/Shared/Net/MsgCenter.cs
using System.Collections.Generic;

namespace Shared.Net
{
	/// <summary>
	/// 消息处理中心
	/// </summary>
	public class MsgCenter
	{
		public delegate ErrorCode GeneralHandler( Google.Protobuf.IMessage message );

		private readonly Dictionary<Protos.MsgID, GeneralHandler> _generalHandlers = new Dictionary<Protos.MsgID, GeneralHandler>();

		public void Register( Protos.MsgID msgID, GeneralHandler handler )
		{
			if ( this._generalHandlers.ContainsKey( msgID ) )
				return;
			this._generalHandlers[msgID] = handler;
		}

		public bool TryGetHandler( Protos.MsgID msgID, out GeneralHandler handler )
		{
			return this._generalHandlers.TryGetValue( msgID, out handler );
		}
	}
}

File: Client/src/Net/MsgCenter.ts
type GeneralHandler = (message:any) => void;

export class MsgCenter {
	private readonly _generalHandlers: Map<number, GeneralHandler>;

	constructor() {
		this._generalHandlers = new Map<number, GeneralHandler>();
	}

	public Register(msgID: number, handler: GeneralHandler): void {
		if (this._generalHandlers.has(msgID))
			return;
		this._generalHandlers[msgID] = handler;
	}

	public Unregister(msgID: number, handler: GeneralHandler): boolean {
		return this._generalHandlers.delete(msgID);
	}

	public GetHandler(msgID: number): GeneralHandler {
		return this._generalHandlers.get(msgID);
	}
}